konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineProfiler.java

```diff
@@ -0,0 +1,213 @@
+            this.currentLog = nextFile;
+        }
+    }
+
+    public PipelineProfiler(ProfilerConfig profilerConfig) {
+        this.profilerConfig = profilerConfig;
+        this.currentLog = profilerConfig.getOutputFile();
+        try {
+            this.writer = new BufferedWriter(new FileWriter(this.currentLog.toString(), true));
+            this.writer.write("[");     //JSON array open (array close is optional for Chrome profiler format)
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        this.json = new ObjectMapper();
+        this.pid = getProcessId();
+        this.tid = Thread.currentThread().getId();
+
+        //Set up a queue so file access doesn't add latency to the execution thread
+        writeQueue =
+                new LinkedBlockingDeque<>();
+        fileWritingThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    runHelper();
+                } catch (Throwable t) {
+                    log.error("Error when attempting to write results to file", t);
+                }
+            }
+
+            public void runHelper() throws Exception {
+                while (true) {
+                    fileSizeGuard();
+                    TraceEvent te = writeQueue.take();    //Blocking
+                    writing.set(true);
+                    try {
+                        String j = json.writeValueAsString(te);
+                        writer.append(j);
+                        writer.append(",\n");
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    } finally {
+                        writing.set(false);
+                    }
+                }
+            }
+        });
+        fileWritingThread.setDaemon(true);
+        fileWritingThread.start();
+    }
+
+    @Override
+    public boolean profilerEnabled() {
+        return true;
+    }
+
+    @Override
+    public void eventStart(String key) {
+        logActive = true;
+        startTime = System.nanoTime() / 1000;
+
+        TraceEvent event = TraceEvent.builder()
+                .name(key)
+                .cat("START")
+                .ts(startTime)
+                .ph(TraceEvent.EventType.B)
+                .pid(this.pid)
+                .tid(this.tid)
+                .build();
+
+        writeQueue.add(event);
+    }
+
+    @Override
+    public void eventEnd(String key) {
+        if (logActive) {
+            while ((!writeQueue.isEmpty() || writing.get()) && fileWritingThread.isAlive()) {
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+            try {
+                writer.flush();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
```

Comment:
This can't be in the logging method. The whole point of having a separate thread for file is to make sure we don't add _any_ overhead to the main thread. If you properly flush and close the writer in fileSizeGuard() this shouldn't be necessary either.